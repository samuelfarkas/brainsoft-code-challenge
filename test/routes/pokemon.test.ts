import { build } from "../helper";

const app = build();

describe("pokemon", () => {
  test("pokemon listing - cursor pagination", async () => {
    const res = await app.inject({
      url: "/pokemon?first=5&cursor=0",
    });
    const payload = res.json();
    expect(res.statusCode).toEqual(200);
    expect(payload.items.length).toEqual(5);
    expect(payload).toEqual(
      expect.objectContaining({
        items: Array(5)
          .fill(null)
          .map((_, index) =>
            // skip check for optional relations due skipped seed control in tests
            expect.objectContaining({
              id: index + 1,
              catalogId: expect.any(String),
              fleeRate: expect.any(Number),
              maxCP: expect.any(Number),
              maxHP: expect.any(Number),
              rarity: expect.any(String),
              classification: {
                id: expect.any(Number),
                name: expect.any(String),
              },
              types: expect.arrayContaining([
                {
                  id: expect.any(Number),
                  name: expect.any(String),
                },
              ]),
              weaknesses: expect.arrayContaining([
                {
                  id: expect.any(Number),
                  name: expect.any(String),
                },
              ]),
              resistant: expect.arrayContaining([
                {
                  id: expect.any(Number),
                  name: expect.any(String),
                },
              ]),
              weightKg: {
                maximum: expect.any(Number),
                minimum: expect.any(Number),
              },
              heightMeters: {
                maximum: expect.any(Number),
                minimum: expect.any(Number),
              },
            }),
          ),
        totalCount: expect.any(Number),
        hasNextPage: true,
        hasPrevPage: false,
      }),
    );
  });

  test("pokemon listing - cursor pagination - second page", async () => {
    const res = await app.inject({
      url: "/pokemon?first=5&cursor=5",
    });
    const payload = res.json();
    expect(res.statusCode).toEqual(200);
    expect(payload.items.length).toEqual(5);
    expect(payload).toEqual(
      expect.objectContaining({
        items: Array(5)
          .fill(null)
          .map((_, index) =>
            expect.objectContaining({
              id: index + 6,
            }),
          ),
        totalCount: expect.any(Number),
        hasNextPage: true,
        hasPrevPage: true,
      }),
    );
  });

  test("pokemon listing - cursor pagination exceeded total count", async () => {
    const res = await app.inject({
      url: "/pokemon?first=5&cursor=5",
    });
    const payload = res.json();
    expect(res.statusCode).toEqual(200);
    expect(payload.items.length).toEqual(5);

    const res2 = await app.inject({
      url: `/pokemon?first=5&cursor=${payload.totalCount + 1}`,
    });
    expect(res2.statusCode).toEqual(200);
    const payload2 = res2.json();
    expect(payload2.items.length).toEqual(0);
  });

  test("pokemon listing - cursor pagination - type", async () => {
    const res = await app.inject({
      url: "/pokemon?first=5&type=10",
    });
    const payload: { items: unknown[] } = res.json();
    expect(res.statusCode).toEqual(200);
    expect(payload.items.length).toBeGreaterThan(0);
    expect(payload).toEqual(
      expect.objectContaining({
        items: payload.items.map(() =>
          expect.objectContaining({
            id: expect.any(Number),
            types: expect.arrayContaining([
              expect.objectContaining({
                id: 10,
              }),
            ]),
          }),
        ),
        totalCount: expect.any(Number),
      }),
    );
  });

  test("pokemon listing - cursor pagination - search", async () => {
    const res = await app.inject({
      url: "/pokemon?first=5&search=pikachu",
    });
    const payload: { items: unknown[] } = res.json();
    expect(res.statusCode).toEqual(200);
    expect(payload.items.length).toBeGreaterThan(0);
    expect(payload).toEqual(
      expect.objectContaining({
        items: payload.items.map(() =>
          expect.objectContaining({
            id: expect.any(Number),
            name: expect.stringMatching(/pikachu/i),
          }),
        ),
        totalCount: 1,
      }),
    );
  });
});
