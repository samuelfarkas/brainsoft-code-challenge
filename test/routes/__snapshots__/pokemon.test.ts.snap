// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pokemon endpoints pokemon by id should find pokemon by id 1`] = `
{
  "attacks": {
    "fast": [
      {
        "attack": {
          "id": 2,
          "name": "Vine Whip",
          "type": {
            "id": 1,
            "name": "Grass",
          },
        },
        "damage": 7,
      },
      {
        "attack": {
          "id": 6,
          "name": "Razor Leaf",
          "type": {
            "id": 1,
            "name": "Grass",
          },
        },
        "damage": 15,
      },
    ],
    "special": [
      {
        "attack": {
          "id": 5,
          "name": "Sludge Bomb",
          "type": {
            "id": 2,
            "name": "Poison",
          },
        },
        "damage": 55,
      },
      {
        "attack": {
          "id": 7,
          "name": "Solar Beam",
          "type": {
            "id": 1,
            "name": "Grass",
          },
        },
        "damage": 120,
      },
      {
        "attack": {
          "id": 8,
          "name": "Petal Blizzard",
          "type": {
            "id": 1,
            "name": "Grass",
          },
        },
        "damage": 65,
      },
    ],
  },
  "catalogId": "003",
  "classification": {
    "id": 1,
    "name": "Seed Pok√©mon",
  },
  "evolutionRequirements": null,
  "fleeRate": 0.05,
  "heightMeters": {
    "maximum": 2.25,
    "minimum": 1.75,
  },
  "id": 1,
  "maxCP": 2392,
  "maxHP": 2580,
  "name": "Venusaur",
  "rarity": "common",
  "resistant": [
    {
      "id": 4,
      "name": "Water",
    },
    {
      "id": 5,
      "name": "Electric",
    },
    {
      "id": 1,
      "name": "Grass",
    },
    {
      "id": 6,
      "name": "Fighting",
    },
    {
      "id": 7,
      "name": "Fairy",
    },
  ],
  "types": [
    {
      "id": 1,
      "name": "Grass",
    },
    {
      "id": 2,
      "name": "Poison",
    },
  ],
  "weaknesses": [
    {
      "id": 8,
      "name": "Fire",
    },
    {
      "id": 9,
      "name": "Ice",
    },
    {
      "id": 12,
      "name": "Flying",
    },
    {
      "id": 13,
      "name": "Psychic",
    },
  ],
  "weightKg": {
    "maximum": 112.5,
    "minimum": 87.5,
  },
}
`;
